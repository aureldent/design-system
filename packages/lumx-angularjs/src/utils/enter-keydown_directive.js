import { ENTER_KEY_CODE } from '@lumx/core/js/constants';

/////////////////////////////

function EnterKeydownDirective() {
    'ngInject';

    function link(scope, el, attrs) {
        el.on('keydown', (evt) => {
            if (evt.which === ENTER_KEY_CODE) {
                scope.$apply(() => {
                    scope.$eval(attrs.lxEnterKeydown, { $event: evt });
                });

                evt.preventDefault();
            }
        });
    }

    return {
        link,
    };
}

/////////////////////////////

angular.module('lumx.utils.enter-keydown').directive('lxEnterKeydown', EnterKeydownDirective);

/////////////////////////////

export { EnterKeydownDirective };
