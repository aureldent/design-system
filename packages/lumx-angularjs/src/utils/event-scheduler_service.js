function EventSchedulerService($document, LxUtilsService) {
    'ngInject';

    const service = this;

    /////////////////////////////
    //                         //
    //    Private attributes   //
    //                         //
    /////////////////////////////

    /**
     * The handlers.
     *
     * @type {Object}
     */
    const _handlers = {};

    /**
     * The schedule.
     *
     * @type {Object}
     */
    const _schedule = {};

    /////////////////////////////
    //                         //
    //    Private functions    //
    //                         //
    /////////////////////////////

    /**
     * Handle en event.
     *
     * @param {Event} evt The event.
     */
    function _handle(evt) {
        let scheduler;

        for (const [key] of Object.entries(_schedule)) {
            if (key.includes(evt.type)) {
                scheduler = _schedule[key];
            }
        }

        if (angular.isDefined(scheduler)) {
            for (let i = 0, len = scheduler.length; i < len; i++) {
                const handler = scheduler[i];

                if (angular.isDefined(handler) && angular.isDefined(handler.cb) && angular.isFunction(handler.cb)) {
                    handler.cb(evt);

                    if (evt.isPropagationStopped()) {
                        break;
                    }
                }
            }
        }
    }

    /////////////////////////////
    //                         //
    //     Public functions    //
    //                         //
    /////////////////////////////

    /**
     * Register en event.
     *
     * @param  {string}   eventName The event name.
     * @param  {Function} cb        The event callback.
     * @return {string}   The event id.
     */
    function register(eventName, cb) {
        const handler = {
            cb,
            eventName,
        };

        const id = LxUtilsService.generateUUID();
        _handlers[id] = handler;

        if (angular.isUndefined(_schedule[eventName])) {
            _schedule[eventName] = [];

            $document.on(eventName, _handle);
        }

        _schedule[eventName].unshift(_handlers[id]);

        return id;
    }

    /**
     * Unregister en event.
     *
     * @param  {string} id The event id.
     * @return {found}  Whether the event has been found or not.
     */
    function unregister(id) {
        let found = false;
        const handler = _handlers[id];

        if (angular.isDefined(handler) && angular.isDefined(_schedule[handler.eventName])) {
            const index = _schedule[handler.eventName].indexOf(handler);

            if (angular.isDefined(index) && index > -1) {
                _schedule[handler.eventName].splice(index, 1);

                delete _handlers[id];
                found = true;
            }

            if (_schedule[handler.eventName].length === 0) {
                delete _schedule[handler.eventName];

                $document.off(handler.eventName, _handle);
            }
        }

        return found;
    }

    /////////////////////////////

    service.register = register;
    service.unregister = unregister;
}

/////////////////////////////

angular.module('lumx.utils.event-scheduler').service('LxEventSchedulerService', EventSchedulerService);

/////////////////////////////

export { EventSchedulerService };
