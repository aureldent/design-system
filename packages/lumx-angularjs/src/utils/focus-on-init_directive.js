import { COMPONENT_PREFIX, MODULE_NAME } from '@lumx/angularjs/constants/common_constants';

/////////////////////////////

function FocusOnInitDirective($timeout) {
    'ngInject';

    function link(scope, el, attrs) {
        if (angular.isDefined(attrs.lumxFocusOnInit) && attrs.lumxFocusOnInit && !scope.$eval(attrs.lumxFocusOnInit)) {
            return;
        }

        $timeout(() => {
            el.focus();
        });
    }

    return {
        link,
    };
}

/////////////////////////////

angular.module(`${MODULE_NAME}.utils.focus-on-init`).directive(`${COMPONENT_PREFIX}FocusOnInit`, FocusOnInitDirective);

/////////////////////////////

export { FocusOnInitDirective };
