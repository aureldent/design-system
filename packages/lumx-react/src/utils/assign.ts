export const assign = (target: any, ...sources: any[]) => {
    sources.forEach((source) => Object.keys(source).forEach((key) => (target[key] = source[key])));
    return target;
};
