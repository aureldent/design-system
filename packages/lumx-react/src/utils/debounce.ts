export function debounce(func: (...args: any) => any, debounceTime: number) {
    let timeout: any;
    return function debouncedFunction(...args: any[]) {
        clearTimeout(timeout);
        timeout = setTimeout(() => func(...args), debounceTime);
    };
}
