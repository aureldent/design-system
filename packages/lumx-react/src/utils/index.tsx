export { getBasicClass, handleBasicClasses, detectSwipe, onEnterPressed, onEscapePressed } from '@lumx/core/utils';

export * from './calculatePopoverPlacement';

export * from './isInViewPort';

export * from './validation';

export * from './partitionMulti';

export { getRootClassName } from './getRootClassName';

export * from './type';
