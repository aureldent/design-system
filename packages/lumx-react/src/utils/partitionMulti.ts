import { Predicate } from './type';
import reduce from 'lodash/reduce';
import concat from 'lodash/concat';
import dropRight from 'lodash/dropRight';
import partition from 'lodash/partition';
import last from 'lodash/last';

/**
 * Similar to lodash `partition` function but working with multiple predicates.
 *
 * @example
 * const isString = (s) => typeof s === 'string'
 * const isNumber = (s) => typeof s === 'number'
 * const [strings, numbers, others] = partitionMulti(['a', 1, 'b', false], [isString, isNumber])
 * //=> [['a', 'b'], [1], [false]]
 *
 * @param  elements array of elements
 * @param  predicates array of predicates to apply on elements
 * @return partitioned elements by the given predicates
 */
function partitionMulti<T>(elements: T[], predicates: Array<Predicate<T>>): T[][] {
    return reduce(
        predicates,
        (partitioned: T[][], predicate: Predicate<T>) =>
            concat(dropRight(partitioned), partition(last(partitioned), predicate)),
        [elements],
    );
}

export { partitionMulti };
