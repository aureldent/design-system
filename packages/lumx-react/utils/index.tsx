export * from './calculatePopoverPlacement';

export * from './isInViewPort';
