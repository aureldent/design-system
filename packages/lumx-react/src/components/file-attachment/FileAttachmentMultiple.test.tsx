import React, { ReactElement } from 'react';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';

import { ICommonSetup, Wrapper, commonTestsSuite } from '@lumx/react/testing/utils';
import { getBasicClass } from '@lumx/react/utils';

import { mdiFile, mdiFilePdfBox } from '@lumx/icons';
import { Theme } from '..';
import {
    CLASSNAME,
    DEFAULT_PROPS,
    FileAttachmentMultiple,
    FileAttachmentMultipleProps,
} from './FileAttachmentMultiple';

/////////////////////////////

/**
 * Define the overriding properties waited by the `setup` function.
 */
type ISetupProps = Partial<FileAttachmentMultipleProps>;

/**
 * Defines what the `setup` function will return.
 */
interface ISetup extends ICommonSetup {
    props: ISetupProps;

    /**
     * The wrapper of the file attachment.
     */
    wrapper: Wrapper;
    /**
     * The ListItem components.
     */
    listItems: Wrapper;
}

/////////////////////////////

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 *
 * @param  props  The props to use to override the default props of the component.
 * @param  [shallowRendering=true] Indicates if we want to do a shallow or a full rendering.
 * @return An object with the props, the component wrapper and some shortcut to some element inside of the component.
 */
const setup = (props: ISetupProps = {}, shallowRendering: boolean = true): ISetup => {
    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;

    // @ts-ignore
    const wrapper: Wrapper = renderer(<FileAttachmentMultiple {...props} />);

    return {
        listItems: wrapper.find('ListItem'),
        props,
        wrapper,
    };
};

const files = [
    {
        thumbnail: { image: 'https://loremflickr.com/320/240' },
        url: 'https://cdn.com/api/pdf/presentation-2020.pdf',
    },
    {
        icon: { icon: mdiFilePdfBox },
        url: 'https://cdn.com/api/pdf/presentation-2020-2.pdf',
    },
    {
        url: 'https://cdn.com/api/pdf/presentation-2020-3.pdf',
    },
];

describe(`<${FileAttachmentMultiple.displayName}>`, () => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', () => {
        // Here is an example of a basic rendering check, with snapshot.

        it('should render correctly', () => {
            const { wrapper } = setup({
                files,
            });
            expect(wrapper).toMatchSnapshot();

            expect(wrapper).toExist();
            expect(wrapper).toHaveClassName(CLASSNAME);
        });
    });

    /////////////////////////////

    // 2. Test defaultProps value and important props custom values.
    describe('Props', () => {
        // Here are some examples of basic props check.

        it('should use default props', () => {
            const { wrapper } = setup({ files });
            Object.keys(DEFAULT_PROPS).forEach((prop: string) => {
                expect(wrapper).toHaveClassName(
                    getBasicClass({ prefix: CLASSNAME, type: prop, value: DEFAULT_PROPS[prop] }),
                );
            });
        });

        it('should pass the theme to the Icons', () => {
            const expectedTheme = Theme.dark;
            const { listItems } = setup({ files, theme: expectedTheme }, false);
            listItems.find('Icon').forEach((icon: Wrapper) => {
                expect(icon).toHaveProp('theme', expectedTheme);
            });
        });

        it('should pass the corresponding classNames to the Icons', () => {
            const expectedClassName = ['test-1', 'test-2'];
            const { listItems } = setup(
                {
                    files: [
                        {
                            icon: { icon: mdiFile, className: 'test-1' },
                            url: 'https://sometest.url/path/to/file/test1.pdf',
                        },
                        {
                            icon: { icon: mdiFile, className: 'test-2' },
                            url: 'https://sometest.url/path/to/file/test2.pdf',
                        },
                    ],
                },
                false,
            );
            listItems.forEach((listItem: Wrapper, index: number) => {
                expect(listItem.find('Icon')).toHaveClassName(expectedClassName[index]);
            });
        });

        it('should display the files names correctly', () => {
            const expectedFileNames = ['test.pdf', 'test2.pdf'];
            const { listItems } = setup({
                files: [
                    { url: 'https://sometest.url/path/to/file/test.pdf' },
                    { url: 'https://sometest.url/path/to/file/test2.pdf' },
                ],
            });
            listItems.forEach((listItem: Wrapper, index: number) => {
                expect(listItem.find('Link')).toHaveProp('children', expectedFileNames[index]);
            });
        });
    });

    /////////////////////////////

    // 3. Test events.
    describe('Events', () => {
        it('should open the corresponding link in a new tab when you click on Thumbnail or Icon', () => {
            const { listItems } = setup(
                {
                    files,
                },
                false,
            );
            window.open = jest.fn();
            listItems.forEach((listItem: Wrapper, index: number) => {
                const thumbnail = listItem.find('Thumbnail');
                const beforeElement = thumbnail.exists() ? thumbnail : listItem.find('Icon');

                if (beforeElement.exists()) {
                    beforeElement.simulate('click');

                    expect(window.open).toHaveBeenCalledWith(files[index].url, '_blank');
                }
            });
        });
    });
    /////////////////////////////

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', () => {
        it('should display Thumbnail if thumbnail is defined and Icon if not', () => {
            const { wrapper } = setup({ files }, false);

            expect(wrapper.find('Icon').length).toBe(2);
            expect(wrapper.find('Thumbnail').length).toBe(1);
        });
    });

    /////////////////////////////

    // 5. Test state.
    describe('State', () => {
        // Nothing to do here.
    });

    /////////////////////////////

    // Common tests suite.
    commonTestsSuite(setup, {}, { className: CLASSNAME });
});
