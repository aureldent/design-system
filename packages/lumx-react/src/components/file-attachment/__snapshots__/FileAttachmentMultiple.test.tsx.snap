// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FileAttachmentMultiple> Snapshots and structure should render correctly 1`] = `
<List
  className="lumx-file-attachment-multiple lumx-file-attachment-multiple--theme-light"
>
  <ListItem
    before={
      <Thumbnail
        aspectRatio="square"
        image="https://loremflickr.com/320/240"
        onClick={[Function]}
        role="link"
        size="m"
        tabIndex={0}
        variant="rounded"
      />
    }
    className="lumx-file-attachment-multiple__file"
  >
    <Link
      className="lumx-file-attachment-multiple__link"
      color="blue"
      colorVariant="N"
      href="https://cdn.com/api/pdf/presentation-2020.pdf"
      target="_blank"
    >
      presentation-2020.pdf
    </Link>
  </ListItem>
  <ListDivider
    className="lumx-file-attachment-multiple__divider"
  />
  <ListItem
    before={
      <Icon
        className="lumx-file-attachment-multiple__file-icon"
        color="dark"
        icon="M19,3A2,2 0 0,1 21,5V19A2,2 0 0,1 19,21H5C3.89,21 3,20.1 3,19V5C3,3.89 3.89,3 5,3H19M10.59,10.08C10.57,10.13 10.3,11.84 8.5,14.77C8.5,14.77 5,16.58 5.83,17.94C6.5,19 8.15,17.9 9.56,15.27C9.56,15.27 11.38,14.63 13.79,14.45C13.79,14.45 17.65,16.19 18.17,14.34C18.69,12.5 15.12,12.9 14.5,13.09C14.5,13.09 12.46,11.75 12,9.89C12,9.89 13.13,5.95 11.38,6C9.63,6.05 10.29,9.12 10.59,10.08M11.4,11.13C11.43,11.13 11.87,12.33 13.29,13.58C13.29,13.58 10.96,14.04 9.9,14.5C9.9,14.5 10.9,12.75 11.4,11.13M15.32,13.84C15.9,13.69 17.64,14 17.58,14.32C17.5,14.65 15.32,13.84 15.32,13.84M8.26,15.7C7.73,16.91 6.83,17.68 6.6,17.67C6.37,17.66 7.3,16.07 8.26,15.7M11.4,8.76C11.39,8.71 11.03,6.57 11.4,6.61C11.94,6.67 11.4,8.71 11.4,8.76Z"
        onClick={[Function]}
        role="link"
        size="m"
        tabIndex={0}
        theme="light"
      />
    }
    className="lumx-file-attachment-multiple__file"
  >
    <Link
      className="lumx-file-attachment-multiple__link"
      color="blue"
      colorVariant="N"
      href="https://cdn.com/api/pdf/presentation-2020-2.pdf"
      target="_blank"
    >
      presentation-2020-2.pdf
    </Link>
  </ListItem>
  <ListDivider
    className="lumx-file-attachment-multiple__divider"
  />
  <ListItem
    before={
      <Icon
        className="lumx-file-attachment-multiple__file-icon"
        color="dark"
        icon="M13,9V3.5L18.5,9M6,2C4.89,2 4,2.89 4,4V20A2,2 0 0,0 6,22H18A2,2 0 0,0 20,20V8L14,2H6Z"
        onClick={[Function]}
        role="link"
        size="m"
        tabIndex={0}
        theme="light"
      />
    }
    className="lumx-file-attachment-multiple__file"
  >
    <Link
      className="lumx-file-attachment-multiple__link"
      color="blue"
      colorVariant="N"
      href="https://cdn.com/api/pdf/presentation-2020-3.pdf"
      target="_blank"
    >
      presentation-2020-3.pdf
    </Link>
  </ListItem>
</List>
`;
